### ISSUE TITLE
SIGINT masking and custom handling crashes with --enable-tsan on x86 macOS

### ISSUE THREAD

**Issue by OlivierNicole** [ocaml-multicore/ocaml-tsan#33]
Consider the following C stub:

```c


#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>

int r = 0;

void handler(int signo)
{
	r = 1;
}

int main1(int unused)
{
	(void) unused;
	// Mask SIGINT
	sigset_t set;
	sigemptyset(&set);
	sigaddset(&set, SIGINT);
	int ret = sigprocmask(SIG_BLOCK, &set, NULL);
	if (ret != 0)
		exit(2);

	// Install handler for SIGINT
	struct sigaction act = { 0 };
	sigemptyset(&act.sa_mask);
	act.sa_flags = SA_ONSTACK;
	act.sa_handler = &handler;
	if (sigaction(SIGINT, &act, NULL) != 0)
		exit(3);

	// Send SIGINT to self
	ret = kill(getpid(), SIGINT);
	if (ret != 0)
		exit(2);

	// Unmask SIGINT
	sigemptyset(&set);
	sigaddset(&set, SIGINT);
	ret = sigprocmask(SIG_UNBLOCK, &set, NULL);
	if (ret != 0)
		exit(2);

	// Should print "r = 1"
	printf("r = %d\n", r);
	return EXIT_SUCCESS;
}
```

and the following calling OCaml code:

```ocaml

external main1 : unit -> unit = "main1"

let () =
        main1 ()
```

Building it with a TSan-enabled compiler using `ocamlopt stub.c test.ml` and running it will result in uninterruptible hang. @fabbing and I could get the following LLDB backtrace:

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x7bd000020000)
  * frame #0: 0x00000001007e6ef7 libclang_rt.tsan_osx_dynamic.dylib`__sanitizer::internal_memcpy(void*, void const*, unsigned long) + 167
    frame #1: 0x0000000100786a6a libclang_rt.tsan_osx_dynamic.dylib`sighandler(int, __sanitizer::__sanitizer_siginfo*, void*) + 474
    frame #2: 0x00007ff80a4c3dfd libsystem_platform.dylib`_sigtramp + 29
    frame #3: 0x00007ff80a474343 libsystem_kernel.dylib`sigprocmask + 11
    frame #4: 0x0000000100794225 libclang_rt.tsan_osx_dynamic.dylib`wrap_sigprocmask + 229
    frame #5: 0x00000001000078fc test`main1 + 172
    frame #6: 0x0000000100054fb4 test`caml_c_call + 288
    frame #7: 0x00000001000020f9 test`camlTest.entry + 41
    frame #8: 0x0000000100001bdb test`caml_program + 75
    frame #9: 0x00000001000550b0 test`caml_start_program + 124
    frame #10: 0x000000010005456e test`caml_startup_common(argv=0x00007ff7bfeff7b8, pooling=<unavailable>) at startup_nat.c:132:9 [opt]
    frame #11: 0x000000010005462c test`caml_main [inlined] caml_startup_exn(argv=0x00007ff7bfeff7b8) at startup_nat.c:139:10 [opt]
    frame #12: 0x0000000100054622 test`caml_main [inlined] caml_startup(argv=0x00007ff7bfeff7b8) at startup_nat.c:144:15 [opt]
    frame #13: 0x0000000100054622 test`caml_main(argv=0x00007ff7bfeff7b8) at startup_nat.c:151:3 [opt]
    frame #14: 0x00000001000367fa test`main(argc=<unavailable>, argv=0x00007ff7bfeff7b8) at main.c:37:3 [opt]
    frame #15: 0x00000001000ad51e dyld`start + 462
```

I know that the TSan runtime handles all signals and passes them to the program. Maybe some form of conflict with OCaml’s signal handling? Interestingly, the bug disappears if you remove the `SA_ONSTACK` flag above. Maybe some subtle difference between macOS and Linux regarding `sigaltstack` or something. (Note: setting `TSAN_OPTIONS=use_sigaltstack=false` does not fix the crash.)

**Comment by OlivierNicole** [comment]
This comes from investigating the failure of `tests/lib-unix/kill/unix_kill.ml` on intel macOS.

**Comment by ghost** [comment]
TL;DR: This is a bug in the TSAN runtime, you can probably reproduce it with the C chunk alone and without the need for any OCaml code to be involved. [edit: you would need to allocate a stack and call `sigaltstack` in the C code, though]

Long version:
The TSAN runtime attempts to make a copy of the signal handler arguments (this is the `internal_memcpy` call in the traceback). But when compiled under MacOS, it will use a default definition of 128 bytes for `siginfo_t` (which in TSAN is called `__sanitizer_siginfo`). However, the Darwin/Xnu/whatever-its-name-those-days uses a shorter struct of roughly 104 bytes (see `bsd/sys/signal.h`). Thus, the `internal_memcpy` call copies beyond the end of the object. This does not show when NOT using `SA_ONSTACK`, because there will likely be existing data on the stack after the `siginfo_t`, so the overrun goes unnoticed. But when using a dedicated stack, it will be (almost) the first thing pushed to the stack, and accesses beyond it will hit an unmapped virtual page.

So until TSAN is fixed to pick a correct definition for `siginfo_t` on MacOS, there is nothing you can do to prevent this.

[I would work on this but I do not have access to an amd64-based MacOS system]

**Comment by kayceesrk** [comment]
> [I would work on this but I do not have access to an amd64-based MacOS system]

I'll find out if we have machines at Tarides. Will ping you on Slack @dustanddreams.

**Comment by OlivierNicole** [comment]
Amazing, I already said this in private but thanks again for the find @dustanddreams.

**Comment by gasche** [comment]
I followed this from the sideline. Impressive! @dustanddreams, are you planning to send a bug report to the TSAN people? That would be nice.

**Comment by ghost** [comment]
> I followed this from the sideline. Impressive! @dustanddreams, are you planning to send a bug report to the TSAN people? That would be nice.

I intend to submit a bug report and a fix, but I would like to be able to confirm this and test the fix on an x86 macos system first, so this will need to wait a bit.

**Comment by ghost** [comment]
I have confirmed that with the testcase modified as:
```
#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>

int r = 0;

void handler(int signo)
{
        r = 1;
}

int main()
{
        void *altstack = malloc(2 * MINSIGSTKSZ);
        if (altstack == NULL)
                perror("malloc");

        stack_t stk = { .ss_sp = altstack, .ss_size = 2 * MINSIGSTKSZ,
            .ss_flags = 0 };
        if (sigaltstack(&stk, NULL) != 0)
                perror("sigaltstack");

        // Mask SIGINT
        sigset_t set;
        sigemptyset(&set);
        sigaddset(&set, SIGINT);
        int ret = sigprocmask(SIG_BLOCK, &set, NULL);
        if (ret != 0)
                perror("sigprocmask");

        // Install handler for SIGINT
        struct sigaction act = { 0 };
        sigemptyset(&act.sa_mask);
        act.sa_flags = SA_ONSTACK;
        act.sa_handler = &handler;
        if (sigaction(SIGINT, &act, NULL) != 0)
                perror("sigaction");

        // Send SIGINT to self
        ret = kill(getpid(), SIGINT);
        if (ret != 0)
                perror("kill");

        // Unmask SIGINT
        sigemptyset(&set);
        sigaddset(&set, SIGINT);
        ret = sigprocmask(SIG_UNBLOCK, &set, NULL);
        if (ret != 0)
                perror("sigprocmask");

        // Should print "r = 1"
        printf("r = %d\n", r);
        return EXIT_SUCCESS;
}
```
We get the following results:
```
$ cc sig.c
$ ./a.out
r = 1
$ cc -fsanitize=thread sig.c
$ ./a.out
a.out(46709,0x7ff84f5c8640) malloc: nano zone abandoned due to inability to reserve vm space.
[then process is stuck, can't be interrupted with ^C, needs to be killed with kill -9
 from another terminal]
```

**Comment by ghost** [comment]
Reported to LLVM, with a possible fix, [here](https://github.com/llvm/llvm-project/issues/63824).
